src/humanloop/otel/exporter.py

```diff
@@ -0,0 +1,272 @@
+import json
+import logging
+import typing
+from queue import Empty as EmptyQueue
+from queue import Queue
+from threading import Thread
+from typing import Any, Optional
+
+from opentelemetry import trace
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
+
+from humanloop.core import ApiError as HumanloopApiError
+from humanloop.eval_utils.context import EVALUATION_CONTEXT, EvaluationContext
+from humanloop.otel import TRACE_FLOW_CONTEXT, FlowContext
+from humanloop.otel.constants import HL_FILE_KEY, HL_FILE_TYPE_KEY, HL_LOG_KEY, HL_PATH_KEY
+from humanloop.otel.helpers import is_humanloop_span, read_from_opentelemetry_span
+from humanloop.requests.flow_kernel_request import FlowKernelRequestParams
+from humanloop.requests.prompt_kernel_request import PromptKernelRequestParams
+
+if typing.TYPE_CHECKING:
+    from humanloop.base_client import BaseHumanloop
+
+
+logger = logging.getLogger("humanloop.sdk")
+
+
+class HumanloopSpanExporter(SpanExporter):
+    """Upload Spans created by SDK decorators to Humanloop.
+
+    Spans not created by Humanloop SDK decorators will be ignored.
+    """
+
+    DEFAULT_NUMBER_THREADS = 4
+
+    def __init__(
+        self,
+        client: "BaseHumanloop",
+        worker_threads: Optional[int] = None,
+    ) -> None:
+        """Upload Spans created by SDK decorators to Humanloop.
+
+        Spans not created by Humanloop SDK decorators will be ignored.
+        """
+        super().__init__()
+        self._client = client
+        # Uploaded spans translate to a Log on Humanloop. The IDs are required to link Logs in a Flow Trace
+        self._span_id_to_uploaded_log_id: dict[int, Optional[str]] = {}
+        # Work queue for the threads uploading the spans
+        self._upload_queue: Queue = Queue()
+        # Worker threads to export the spans
+        self._threads: list[Thread] = [
+            Thread(
+                target=self._do_work,
+                daemon=True,
+            )
+            for _ in range(worker_threads or self.DEFAULT_NUMBER_THREADS)
+        ]
+        # Signals threads no more work will arrive and
+        # they should wind down if the queue is empty
+        self._shutdown: bool = False
+        for thread in self._threads:
+            thread.start()
+
+    def export(self, spans: trace.Sequence[ReadableSpan]) -> SpanExportResult:
+        if not self._shutdown:
+            try:
+                evaluation_context = EVALUATION_CONTEXT.get()
+            except LookupError:
+                evaluation_context = None
+            for span in spans:
+                if is_humanloop_span(span):
+                    # The thread doing the logging is different than the
+                    # thread spawned by eval_utils.run.run_eval. Need
+                    # to pass the EvaluationContext to the thread doing
+                    # the logging
```

Comment:
i don't understand the implication of this comment.
is it trying to explain why `evaluation_context` is in the line below? (if so, it should explicitly say that's what it's talking about - something like "We need to include `evaluation_context` in the item we put into `_upload_queue` as the thread doing the logging ... so it can ... .")

Reply:
ack changed